### PR TITLE
WIP: Vault Agent: Add configuration for maximum connection attempts

### PR DESCRIPTION
**Pull request by catsby**
WIP

Vault Agent can be configured to attempt to authenticate with a Vault cluster that is not responsive. When this occurs, the goroutine running the Auth Method can enter an infinite loop of attempting to connect and receiving tcp dial errors like so:

```
2020-01-10T16:33:13.848-0600 [INFO]  auth.handler: authenticating
2020-01-10T16:33:13.849-0600 [ERROR] auth.handler: error authenticating: error="Put http://127.0.0.1:8200/v1/auth/approle/login: dial tcp 127.0.0.1:8200: connect: connection refused" backoff=2.496237872
```

This repeats until either Vault becomes responsive or the Agent process is terminated. Note that this is different than a failed auth attempt or responses from a sealed Vault; the dial error occurs when no connection can be made at the tcp layer. Because of that fact the existing `VAULT_CLIENT_TIMEOUT` and `VAULT_MAX_RETRIES` configuration options are never even considered, and Agent attempts to authenticate indefinitely. 

This pull request adds an additional configuration, `max-connection-attempts` which is specific to Vault Agent. This option can limit the number of failed connection attempts before exiting the Agent process. The default is `0`, which indicates "unlimited", so that the current behavior is unchanged and this change is backwards compatible. 

The maximum connection attempts can be set with the `-max-connection-attempts` command line flag, the `VAULT_MAX_CONNECTION_ATTEMPTS` environment variable, or the `max_connection_attempts` configuration variable, in that order of precedence. 

------

Remaining items:

- [ ] Clean up / add code comments
- [ ] Agent web documentation

### CODE REVIEW

File: command/agent.go [hunk 1]
```diff
@@ -116,6 +120,15 @@ func (c *AgentCommand) Flags() *FlagSets {
 			"\"trace\", \"debug\", \"info\", \"warn\", and \"err\".",
 	})
 
+	f.IntVar(&IntVar{
+		Name:    "max-connection-attempts",
+		Target:  &c.flagMaxConnectionAttempts,
+		Default: 0,
+		EnvVar:  "VAULT_MAX_CONNECTION_ATTEMPTS",
```
Comment (kalafut) [inline command/agent.go]: Why is connection attempts preferred over a timeout as a configuration?  As a user I know was "60s" means, but I don't know how long a connection attempt takes.

Comment (catsby) [reply command/agent.go]: That's a fair question. The original request for this was for something akin to [`VAULT_MAX_RETRIES`][max_retries]. A timeout could likely work just as well


[max_retries]: https://www.vaultproject.io/docs/commands/index.html#vault_max_retries

Comment (jasonodonnell) [reply command/agent.go]: Timeout works for me!

File: command/agent/auth/auth.go [hunk 2]
```diff
@@ -71,18 +71,36 @@ func backoffOrQuit(ctx context.Context, backoff time.Duration) {
...
+	var connTimer *time.Timer
 	ah.logger.Info("starting auth handler")
 	defer func() {
 		am.Shutdown()
 		close(ah.OutputCh)
 		close(ah.DoneCh)
 		close(ah.TemplateTokenCh)
 		ah.logger.Info("auth handler stopped")
+		if connTimer != nil {
+			if !connTimer.Stop() {
```
Comment (kalafut) [inline command/agent/auth/auth.go]: I see this example in the Go docs but I don't really understand it, and it feels like we could hang in the channel receive?

**Pull request closed without merging**